cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillPreparer.java

```diff
@@ -166,13 +169,55 @@
     this.twillSpec = spec;
     this.resourcePrefix = resourcePrefix;
     this.extraLabels = extraLabels;
+    dependentRunnableNames = new HashSet<>();
   }
 
+  /**
+   * A call overwrites the values provided in the previous call.
+   */
   @Override
-  public KubeTwillPreparer withStatefulRunnable(String runnableName,
+  public DependentTwillPreparer withDependentRunnables(String mainRunnableName, String... dependentRunnableName) {
+    if (mainRunnableName == null || dependentRunnableName == null || dependentRunnableName.length == 0) {
+      throw new IllegalArgumentException("Main or dependent runnable names cannot be empty");
+    }
+
+    if (twillSpec.getRunnables().get(mainRunnableName) == null) {
+      throw new IllegalArgumentException(String.format("Main runnable name %s has not been set", mainRunnableName));
+    }
+
+    for (String name : dependentRunnableName) {
+      if (twillSpec.getRunnables().get(name) == null) {
+        throw new IllegalArgumentException(String.format("Runnable name %s has not been set", name));
+      }
+    }
+
+    Set<String> tmpSet = new HashSet<>(Arrays.asList(dependentRunnableName));
+    for (RuntimeSpecification spec : this.twillSpec.getRunnables().values()) {
+      tmpSet.remove(spec.getName());
+    }
+    if (tmpSet.size() > 0) {
+      throw new IllegalArgumentException(String.format("Specify dependency for runnables %s", tmpSet.toString()));
+    }
+
+    this.mainRunnableName = mainRunnableName;
```

Comment:
Should also check the `mainRunnableName` is not in the `dependentRunnableNames`.